### ISSUE TITLE
Permissions and ownership states always return as "changed"

### ISSUE THREAD

**Issue by thecosmicfrog** [saltstack/salt#26146]
I'm not sure if this is a bug or if it's expected behavior. I haven't been able to find anything online about this.

Basic summary: Whenever I run a **highstate** or a **state.sls** from my Master to my Minions, any states that manage permissions or ownership always return as "changed" instead of just "succeeded", regardless of whether or not anything changed.

If this is expected behavior, why so? If I see something "changed" by Salt, I'd really like to know what that thing is. But, if it's just the permissions of /etc/myfile.conf being "changed" (_eh-em_) from 640 to 640, that doesn't mean anything to me as an administrator. If anything, it's likely to make me miss **actual** changes.

salt-master and salt-minion versions are **2015.5.3** across the board.
#### Example below:

```
-
salt-master:~# salt '*' state.highstate

[...]

\----------
          ID: icinga2_api_symlink
    Function: file.symlink
        Name: /etc/icinga2/features-enabled/api.conf
      Result: True
     Comment: Set ownership of symlink /etc/icinga2/features-enabled/api.conf to root:root
     Started: 13:42:07.933006
    Duration: 3.823 ms
     Changes:
              ----------
              ownership:
                  root:root
\----------
          ID: icinga2_conf.d
    Function: file.recurse
        Name: /etc/icinga2/conf.d
      Result: True
     Comment: Recursively updated /etc/icinga2/conf.d
     Started: 13:42:07.937026
    Duration: 1618.278 ms
     Changes:
              ----------
              /etc/icinga2/conf.d/apt.conf:
                  ----------
                  mode:
                      0644
              /etc/icinga2/conf.d/commands.conf:
                  ----------
                  mode:
                      0644
              /etc/icinga2/conf.d/hosts.conf:
                  ----------
                  mode:
                      0644
              /etc/icinga2/conf.d/services.conf:
                  ----------
                  mode:
                      0644
              /etc/icinga2/conf.d/users.conf:
                  ----------
                  mode:
                      0644
              /etc/icinga2/conf.d/yum.conf:
                  ----------
                  mode:
                      0644
\----------
          ID: icinga2_constants
    Function: file.managed
        Name: /etc/icinga2/constants.conf
      Result: True
     Comment: File /etc/icinga2/constants.conf updated
     Started: 13:42:09.560539
    Duration: 123.594 ms
     Changes:
              ----------
              mode:
                  0644
\----------

[...]

Summary
\------------
Succeeded: 8 (changed=5)
Failed:    0
\------------
Total states run:     8
```

**Comment by jfindlay** [comment]
@thecosmicfrog, thanks for the report.  The file.managed state function uses the file.check_perms exec function to check whether changes are needed.  It looks like something may be broken either there or in the use of this function in file.managed.

**Comment by thecosmicfrog** [comment]
@jfindlay Thanks for the response. No problem at all. Glad to help. I've noticed this for a while now. Sorry for the vague "while", but definitely since mid-2014 as far as I can tell. In other words, since I first started using Salt in production. That's why I'm asking if it's expected behavior!

Is there anything I can do to help test/debug?

**Comment by jfindlay** [comment]
I think we have the external features of this problem well enough defined.  If you have some time and desire to look over the source code, that would be great. :-)

**Comment by thecosmicfrog** [comment]
@jfindlay Exactly what I'm doing now!

**Comment by thecosmicfrog** [comment]
Hmmm... something funky does indeed look like it's happening with the file.check_perms module. As below:

-

root@salt-master:~# salt 'server1' file.check_perms /tmp/file.txt '{}' root root 640
server1:
    |_
      ----------
      changes:
          ----------
      comment:
      name:
          /tmp/file.txt
      result:
          True
    |_
      ----------
      lgroup:
          root
      lmode:
          0640
      luser:
          root

-

Followed right after with the exact same:

root@salt-master:~# salt 'server1' file.check_perms /tmp/file.txt '{}' root root 640
server1:
    |_
      ----------
      changes:
          ----------
      comment:
      name:
          /tmp/file.txt
      result:
          True
    |_
      ----------
      lgroup:
          root
      lmode:
          0640
      luser:
          root

-
This is continuously reproducible.
-

Note that **/tmp/file.txt** is considered in **changes:** despite the owner/group and permissions not having changed.

Will keep digging...

**Comment by thecosmicfrog** [comment]
Wait, that above post makes no sense. Ignore it. **/tmp/file.txt** is **not** considered in "changes". Must have gone selectively blind yesterday. More digging...

**Comment by thecosmicfrog** [comment]
Hello again. It's almost a year later and I am still having this behaviour with Salt 2015.8.10. Is there any update on this?

**Comment by jfindlay** [comment]
@thecosmicfrog, not that I know of.  Ping @cachedout.

**Comment by ixs** [comment]
Stumbled over this bug as it described an issue I had as well.
Executing a highstate would _always_ modify the permissions on a file triggering a server restart.

I added a bit of debugging with log.info(name, mode, perms['lmode]) to the file module and turns out my issue is not a salt problem but just two permission changes clashing.

Before and after running salt-call state.highstate the file in question was 0644 but looking at the info output from the salt-call the actual mode of the file was 0755. Turns out a different statefile contained a managed.directory() aiming at recursively changing the permissions for a directory tree but forgetting to specify ignore_files.

@thecosmicfrog Maybe that is something happening to you as well?
Line 4318 of file.py would be a good place for a log.debug() entry to see what is going on.
